bindinfo/handle.go

```diff
@@ -169,8 +169,83 @@ func (h *BindHandle) Update(fullLoad bool) (err error) {
 	return nil
 }
 
-// AddBindRecord adds a BindRecord to the storage and BindRecord to the cache.
-func (h *BindHandle) AddBindRecord(sctx sessionctx.Context, record *BindRecord) (err error) {
+// CreateBindRecord creates a BindRecord to the storage and the cache.
+// It replaces all the exists bindings for the same normalized SQL.
+func (h *BindHandle) CreateBindRecord(sctx sessionctx.Context, record *BindRecord) (err error) {
+	err = record.prepareHints(sctx)
+	if err != nil {
+		return err
+	}
+
+	exec, _ := h.sctx.Context.(sqlexec.SQLExecutor)
+	h.sctx.Lock()
+	_, err = exec.ExecuteInternal(context.TODO(), "BEGIN")
+	if err != nil {
+		h.sctx.Unlock()
+		return
+	}
+
+	normalizedSQL := parser.DigestNormalized(record.OriginalSQL)
+	br := h.GetBindRecord(normalizedSQL, record.OriginalSQL, record.Db)
+
+	defer func() {
+		if err != nil {
+			_, err1 := exec.ExecuteInternal(context.TODO(), "ROLLBACK")
+			h.sctx.Unlock()
+			terror.Log(err1)
+			return
+		}
+
+		_, err = exec.ExecuteInternal(context.TODO(), "COMMIT")
+		h.sctx.Unlock()
+		if err != nil {
+			return
+		}
+
+		// update the BindRecord to the cache.
+		if br != nil {
+			// removeBindRecord locks h.bindinfo, so we needn't involve it between lock and unlock.
+			err = h.removeBindRecord(normalizedSQL, br)
+			if err != nil {
+				return
+			}
+		}
+		// Make sure there is only one goroutine writes the cache and use parser.
+		h.bindInfo.Lock()
+		h.appendBindRecord(normalizedSQL, record)
+		h.bindInfo.Unlock()
+	}()
+
+	if br != nil {
+		for _, binding := range br.Bindings {
+			_, err = exec.ExecuteInternal(context.TODO(), h.deleteBindInfoSQL(record.OriginalSQL, record.Db, binding.BindSQL))
```

Comment:
Is it safer to use `logicalDeleteBindInfoSQL` here? Then the user can see all historical bindings.

Reply:
But there is GC thread to delete the `deleted` SQL. Should we still logically delete it to increase the overload of the binding system?

Reply:
Both operations need to write to TiKV, how about physically delete them when creating a new binding to overwrite the old binding or bindings? It does the less work compared with logically delete IMHO.

Reply:
nit: if we delete all the bindings with `DELETE FROM mysql.bind_info WHERE original_sql=%s AND default_db=%s`, it would save some internal SQLs and improve the deletion performance.